### PR TITLE
feat: workforce credentials

### PR DESCRIPTION
**Pull request by bshaffer**
addresses #483 

(note: relevant testing requested in https://github.com/phpspec/prophecy/issues/611)

### CODE REVIEW

**Comment by bshaffer** [comment]
@aeitzman I'd love to get your review on this one as well!

File: src/FetchAuthTokenCache.php [hunk 1]
```diff
@@ -146,9 +146,12 @@ public function signBlob($stringToSign, $forceOpenSsl = false)
...
         }
 
-        // Pass the access token from cache to GCECredentials for signing a blob.
-        // This saves a call to the metadata server when a cached token exists.
-        if ($this->fetcher instanceof Credentials\GCECredentials) {
+        // Pass the access token from cache for credentials that sign blobs
+        // using the IAM API. This saves a call to fetch an access token when a
+        // cached token exists.
+        if ($this->fetcher instanceof Credentials\GCECredentials
+            || $this->fetcher instanceof Credentials\ImpersonatedServiceAccountCredentials
```
Comment (aeitzman) [inline src/FetchAuthTokenCache.php:153]: Is this related to this change?

File: src/FetchAuthTokenCache.php [hunk 2]
```diff
@@ -189,6 +192,15 @@ public function getProjectId(callable $httpHandler = null)
             );
         }
 
+        // Pass the access token from cache for credentials that require an
+        // access token to fetch the project ID. This saves a call to fetch an
+        // access token when a cached token exists.
+        if ($this->fetcher instanceof Credentials\ExternalAccountCredentials) {
```
Comment (aeitzman) [inline src/FetchAuthTokenCache.php:198]: Just double checking since I don't fully understand the caching flow in this library, this call is only used if getProjectId is explicitly called by a user/some other library right? Just want to make sure that the default flow for getting an access token doesn't break if cloudresourcemanager is down.

File: src/FetchAuthTokenCache.php (hunk 1 above)
Comment (bshaffer) [reply src/FetchAuthTokenCache.php:153]: Hmm, no. I am not sure why this would be in here. I probably just realized that it was something that needed fixing, when adding similar logic below for `getProjectId`.

File: src/FetchAuthTokenCache.php (hunk 2 above)
Comment (bshaffer) [reply src/FetchAuthTokenCache.php:198]: Yes that's correct - this is ONLY for calls to `$credentials->getProjectId()`

**Review by aeitzman — APPROVED** [review]

File: tests/Credentials/ExternalAccountCredentialsTest.php [hunk 3]
```diff
@@ -326,15 +318,198 @@ public function testFetchAuthTokenWithImpersonation()
...
+            $this->assertEquals('Bearer some-token', $request->getHeaderLine('authorization'));
+            $body = $this->prophesize(StreamInterface::class);
+            $body->__toString()->willReturn(json_encode(['projectId' => 'test-project-id']));
+
+            $response = $this->prophesize(ResponseInterface::class);
+            $response->getBody()->willReturn($body->reveal());
+            $response->hasHeader('Content-Type')->willReturn(false);
+
+            return $response->reveal();
+        };
```
Comment (yash30201) [inline tests/Credentials/ExternalAccountCredentialsTest.php:409]: Can you help me understand why are we defining this httphandler if the token and eventually the project id would be fetched from the mocked cache?

**Review by yash30201 — COMMENTED** [review]
Just left few nits and comments to double check my understand. Looks good implementation wise.

File: src/Credentials/ExternalAccountCredentials.php [hunk 4]
```diff
@@ -90,14 +101,25 @@ public function __construct(
...
 
         $this->auth = new OAuth2([
             'tokenCredentialUri' => $jsonKey['token_url'],
             'audience' => $jsonKey['audience'],
             'scope' => $scope,
             'subjectTokenType' => $jsonKey['subject_token_type'],
             'subjectTokenFetcher' => self::buildCredentialSource($jsonKey),
+            'additionalOptions' => $this->workforcePoolUserProject
+                ? ['userProject' => $this->workforcePoolUserProject]
+                : [],
```
[the remark this answers is not in the document]
Comment (bshaffer) [reply src/Credentials/ExternalAccountCredentials.php:115]: Why remove the training comma? This is now best practice for arrays (adding a new item will only diff the added line)

File: src/Credentials/ExternalAccountCredentials.php [hunk 5]
```diff
@@ -238,4 +261,68 @@ public function getQuotaProject()
...
+     *        token. **Defaults to** `null`.
+     * @return string|null
+     */
+    public function getProjectId(callable $httpHandler = null, string $accessToken = null)
+    {
+        if (isset($this->projectId)) {
+            return $this->projectId;
+        }
+
+        $projectNumber = $this->getProjectNumber() ?: $this->workforcePoolUserProject;
```
[the remark this answers is not in the document]
Comment (bshaffer) [reply src/Credentials/ExternalAccountCredentials.php:290]: Yes, they are equivalent. As far as I know, one is not preferred over the other (as in this case, the left side would never be undefined)

File: tests/Credentials/ExternalAccountCredentialsTest.php (hunk 3 above)
Comment (bshaffer) [reply tests/Credentials/ExternalAccountCredentialsTest.php:409]: Sure. This method is testing the getProjectId method, which makes an API call using the project number in order to retrieve the project ID. It's also testing that a cached access token is used when that API call is made, instead of retrieving a new one.

So the cache is for the access token, which is needed for the API request to get the project ID. The httpHandler is the HTTP request made when getProjectId is called, which is not cached.

File: src/Credentials/ExternalAccountCredentials.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (bshaffer) [reply src/Credentials/ExternalAccountCredentials.php:290]:
```suggestion
        $projectNumber = $this->getProjectNumber() ?? $this->workforcePoolUserProject;
```

[the remark this answers is not in the document]
Comment (yash30201) [reply src/Credentials/ExternalAccountCredentials.php:290]: > [See here for more info](https://stackoverflow.com/questions/34571330/php-short-ternary-elvis-operator-vs-null-coalescing-operator)

Ohh, thanks for ref. I learnt something new : )   !

> However, I'm happy to change it if you think it's more readable.

It's up to you. Personally I also don't believe in enforcing some way if it's the exact same as the other.

File: tests/Credentials/ExternalAccountCredentialsTest.php (hunk 3 above)
Comment (yash30201) [reply tests/Credentials/ExternalAccountCredentialsTest.php:409]: Oh, now this makes sense. 
I feel adding a small explanation there in comment would go a long way to quickly understand it.
